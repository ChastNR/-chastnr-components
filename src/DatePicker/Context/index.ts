export { DatePickerContextProvider } from './DatePickerContextProvider';
export { useDatePickerContext } from './useDatePickerContext';
