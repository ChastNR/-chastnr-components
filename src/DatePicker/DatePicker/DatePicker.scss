.ctrl__cntr {
  &.dp {
    cursor: pointer;
  }

  .dp__value {
    color: rgba(0, 0, 0, 0.54);
    line-height: 0;
    opacity: 0;
    outline: 0;
    padding: 22px 0 9px;
    position: relative;
    text-overflow: ellipsis;
    transition: opacity 0.25s ease;
    white-space: nowrap;
    width: 100%;
    will-change: opacity;
    word-wrap: break-word;
  }

  .dp__cnt {
    display: flex;
    align-items: center;

    & > div:last-child {
      margin-left: 1em;
    }
  }

  &:focus-within {
    .dp__cntr > div {
      opacity: 1;
      z-index: 100;
      margin-top: 4px;
    }

    &.hasValue {
      .dp__value {
        color: #333;
      }
    }

    .dp__value {
      opacity: 1;
    }
  }

  &.hasValue {
    .dp__value {
      color: #333;
      opacity: 1;
    }
  }

  .dp__cntr > div {
    opacity: 0;
    margin-top: 15px;
    z-index: -1;
    transition: all 0.25s;
  }
}
