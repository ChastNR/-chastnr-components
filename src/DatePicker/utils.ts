const DAYS = ['ВС', 'ПН', 'ВТ', 'СР', 'ЧТ', 'ПТ', 'СБ'];

const MONTHS = [
  'Январь',
  'Февраль',
  'Март',
  'Апрель',
  'Май',
  'Июнь',
  'Июль',
  'Август',
  'Сентябрь',
  'Октябрь',
  'Ноябрь',
  'Декабрь',
];

const CALENDAR_WEEKS = 6;

const zeroPad = (value: number | string, length: number): string =>
  `${value}`.padStart(length, '0');

const todayDate = new Date();

const THIS_YEAR = todayDate.getFullYear();

const THIS_MONTH = todayDate.getMonth() + 1;

const MONTHS_WITH_THIRTY_DAYS = [4, 6, 9, 11];

const getMonthDays = (month = THIS_MONTH, year = THIS_YEAR): number => {
  if (month === 2) {
    const leapYear = year % 4 === 0;
    return leapYear ? 29 : 28;
  }

  return MONTHS_WITH_THIRTY_DAYS.includes(month) ? 30 : 31;
};

const getMonthFirstDay = (month = THIS_MONTH, year = THIS_YEAR): number =>
  +new Date(`${year}-${zeroPad(month, 2)}-01`).getDay() + 1;

const isSameMonth = (date: Date, basedate = new Date()): boolean => {
  const basedateMonth = +basedate.getMonth() + 1;
  const basedateYear = basedate.getFullYear();
  const dateMonth = +date.getMonth() + 1;
  const dateYear = date.getFullYear();
  return +basedateMonth === +dateMonth && +basedateYear === +dateYear;
};

const isSameDay = (date: Date, basedate = new Date()): boolean => {
  const basedateDate = basedate.getDate();
  const dateDate = date.getDate();
  const isSameDateMonth = isSameMonth(date, basedate);
  return +basedateDate === +dateDate && isSameDateMonth;
};

const isWeekend = (date: Date): boolean => {
  const day = date.getDay();
  return day === 6 || day === 0;
};

const getDateISO = (date = new Date()): string =>
  [date.getFullYear(), zeroPad(+date.getMonth() + 1, 2), zeroPad(+date.getDate(), 2)].join('-');

interface PreviousDate {
  month: number;
  year: number;
}

const getPreviousMonth = (month: number, year: number): PreviousDate => ({
  month: month > 1 ? month - 1 : 12,
  year: month > 1 ? year : year - 1,
});

const getNextMonth = (month: number, year: number): PreviousDate => ({
  month: month < 12 ? month + 1 : 1,
  year: month < 12 ? year : year + 1,
});

const createCalendar = (month = THIS_MONTH, year = THIS_YEAR): (string | number)[][] => {
  const monthDays = getMonthDays(month, year);
  const monthFirstDay = getMonthFirstDay(month, year);

  const daysFromPrevMonth = monthFirstDay - 1;
  const daysFromNextMonth = CALENDAR_WEEKS * 7 - (daysFromPrevMonth + monthDays);

  const { month: prevMonth, year: prevMonthYear } = getPreviousMonth(month, year);
  const { month: nextMonth, year: nextMonthYear } = getNextMonth(month, year);

  const prevMonthDays = getMonthDays(prevMonth, prevMonthYear);

  const prevMonthDates = [...new Array(daysFromPrevMonth)].map((_, index) => {
    const day = index + 1 + (prevMonthDays - daysFromPrevMonth);
    return [prevMonthYear, zeroPad(prevMonth, 2), zeroPad(day, 2)];
  });

  const thisMonthDates = [...new Array(monthDays)].map((_, index) => {
    const day = index + 1;
    return [year, zeroPad(month, 2), zeroPad(day, 2)];
  });

  const nextMonthDates = [...new Array(daysFromNextMonth)].map((_, index) => {
    const day = index + 1;
    return [nextMonthYear, zeroPad(nextMonth, 2), zeroPad(day, 2)];
  });

  return [...prevMonthDates, ...thisMonthDates, ...nextMonthDates];
};

const getCalendarTitle = (month: number, year: number): string => {
  const monthName = MONTHS[Math.max(0, Math.min(month - 1, 11))];
  return `${monthName} ${year}`;
};

export {
  DAYS,
  MONTHS,
  CALENDAR_WEEKS,
  zeroPad,
  getMonthDays,
  getMonthFirstDay,
  getCalendarTitle,
  isSameMonth,
  isSameDay,
  isWeekend,
  getDateISO,
  getPreviousMonth,
  getNextMonth,
  createCalendar,
};
