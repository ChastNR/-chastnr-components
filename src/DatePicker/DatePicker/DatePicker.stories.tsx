/* eslint-disable import/no-extraneous-dependencies */

import { ComponentMeta } from '@storybook/react';
import { DatePicker } from './DatePicker';

const componentMeta: ComponentMeta<typeof DatePicker> = {
  title: 'DatePicker',
  component: DatePicker,
  parameters: {
    layout: 'fullscreen',
  },
};

export const DefaultDatePicker = () => <DatePicker label="Test label" onSelect={() => {}} />;

export default componentMeta;
