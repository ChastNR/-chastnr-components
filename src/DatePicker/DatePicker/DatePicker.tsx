import { PropsWithChildren, useCallback, useRef, useState } from 'react';
import dayjs from 'dayjs';
import { Calendar } from '../../base';
import { Control } from '../../Control';
import { createError } from '../../utils';
import { Picker } from '../Picker';

import './DatePicker.scss';

interface IDatePickerProps
  extends PropsWithChildren<{
    className?: string;
    date?: Date;
    disabled?: boolean;
    endDate?: Date;
    errorMessage?: React.ReactNode;
    format?: string;
    label: React.ReactNode;
    onSelect(date: Date): void;
    placeholder?: string;
    requiredMessage?: React.ReactNode;
    startDate?: Date;
  }> {}

export const DatePicker: React.FC<IDatePickerProps> = ({
  className,
  children,
  date,
  disabled,
  endDate,
  errorMessage,
  format = 'DD.MM.YYYY',
  label,
  onSelect,
  placeholder = 'дд.мм.гггг',
  requiredMessage,
  startDate,
}) => {
  const wrapperRef = useRef<HTMLDivElement>(null);

  const [validate, setValidate] = useState(false);

  const handleBlur = useCallback((e: React.FocusEvent<HTMLDivElement>) => {
    if (e.currentTarget.contains(e.relatedTarget as Node)) {
      return;
    }

    setValidate(true);
  }, []);

  const handleSelect = useCallback(
    (value: Date) => {
      setValidate(false);
      wrapperRef.current?.blur();
      onSelect(value);
    },
    [onSelect]
  );

  const error = createError({
    validate,
    requiredMessage,
    value: date,
    errorMessage,
  });

  const value = date ? dayjs(date).format(format) : placeholder;

  return (
    <Control
      className={className}
      content={
        <div className="dp__cnt">
          {children}
          <Calendar />
        </div>
      }
      disabled={disabled}
      error={error}
      hasValue={!!date}
      label={label}
      onBlur={handleBlur}
      ref={wrapperRef}
      required={!!requiredMessage}
    >
      <span className="dp__value">{value}</span>
      <div className="dp__cntr">
        <Picker date={date} endDate={endDate} onSelect={handleSelect} startDate={startDate} />
      </div>
    </Control>
  );
};
