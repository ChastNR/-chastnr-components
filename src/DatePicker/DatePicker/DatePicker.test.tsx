import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";

import DatePicker from "./DatePicker";

const mockOnSelect = jest.fn();

describe("<DatePicker />", () => {
  it("should render", () => {
    const control = shallow(<DatePicker onSelect={mockOnSelect} label="testLabel" />);
    expect(toJson(control)).toMatchSnapshot();
  });
});
