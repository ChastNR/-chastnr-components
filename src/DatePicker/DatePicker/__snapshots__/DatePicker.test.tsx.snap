// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> should render 1`] = `
<ForwardRef
  className="dp"
  content={
    <div
      className="dp__cnt"
    >
      <Memo />
    </div>
  }
  error={null}
  hasValue={false}
  label="testLabel"
  onBlur={[Function]}
  required={false}
>
  <span
    className="dp__value"
  >
    дд.мм.гггг
  </span>
  <div
    className="dp__cntr"
  >
    <Picker
      onSelect={[Function]}
    />
  </div>
</ForwardRef>
`;
